Validate Input Policy and show info on user code caches

shubham3121 opened the pull request:
## Description
Please include a summary of the change, the motivation, and any additional context that will help others understand your PR. If it closes one or more open issues, [please tag them as described here](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [ ] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

shubham3121 committed:
refactor input policy to define a _is_valid method
- define _is_valid method for InputPolicy

shubham3121 committed:
add a check for input policy before executing user custom code

shubham3121 committed:
Merge branch 'dev' into shubham/fix-output-policy

shubham3121 committed:
fix missing context argument to _is_valid for input policy in user code exec

shubham3121 committed:
add a CachedExecutionResult class to represent cache value and store error messages
- prompt a warning message if values are picked from cache

shubham3121 commented:
Caching Warning and propagating error message.

![Screenshot from 2024-03-18 18-46-51](https://github.com/OpenMined/PySyft/assets/11032835/28d7b2e2-ad5b-4af6-a750-2b007a5bd4e9)

shubham3121 committed:
rename a CachedExecutionResult to CachedSyftObject
- make CachedSyftObject a new syft type
- extract ids from kwargs before passing it input policy

shubham3121 commented:
Propagating error when Input policy is not longer valid (e.g. running a function on a different dataset)

![Screenshot from 2024-03-18 19-08-43](https://github.com/OpenMined/PySyft/assets/11032835/f1dff3da-ca9f-4aa5-9f90-7852673ae2da)

shubham3121 committed:
fix propogation of error when input policy doesn't match

shubham3121 committed:
fix variable is_approved scope in _user_execute_code
- check for overrides during input policy validation

teo-milea commented on packages/syft/src/syft/service/action/action_service.py:
```diff
@@ -302,45 +302,34 @@ def _user_code_execute(
             filtered_kwargs = input_policy.filter_kwargs(
                 kwargs=kwargs, context=context, code_item_id=code_item.id
             )
-            if isinstance(filtered_kwargs, SyftError) or filtered_kwargs.is_err():
```
so glad to see we are slowly removing the SyftError type for the internal API!

teo-milea commented on packages/syft/src/syft/protocol/protocol_version.json:
Not sure how the protocol versioning works right now, but why do we have so many changes without changes to objects?

teo-milea reviewed (approved):
Great work!

teo-milea committed:
fix bug with hello-syft nb

shubham3121 committed:
save input_ids as part of ExecutionOutput
- pass input kwargs to apply_output in code.call service
- default to policy input kwargs as execution inputs during accept by depositing results

shubham3121 committed:
check for inputs before loading output from cache in user code execution

shubham3121 committed:
update test to validate results are cached

shubham3121 committed:
add check in notebook to test for different inputs for given approved function

shubham3121 committed:
fix return typing for _is_valid
- add check for input policy is not None

teo-milea committed:
Merge branch 'dev' into shubham/fix-output-policy
